File: src/game/combat.ts
import { ITEMS } from './items.js';
import type { PlayerSession } from './player.js';
import type { LiveMob, World } from './world.js';
import { CLASSES } from './types.js';
import { checkQuestComplete } from './quests.js';

export const SAFE_ZONES = new Set(['Eldermoor']);

export function isSafeZone(world: World, roomId: string): boolean {
  return SAFE_ZONES.has(world.getZone(roomId));
}

export interface CombatResult {
  messages: string[];
  playerDied: boolean;
  mobKilled: boolean;
}

export interface PvpRoundResult {
  attackerMessages: string[];
  defenderMessages: string[];
  defenderDied: boolean;
  attackerDied: boolean;
}

function rollDamage(attack: number, defense: number): number {
  const base = Math.max(1, attack - Math.floor(defense / 2));
  const variance = Math.floor(base * 0.3);
  return base + Math.floor(Math.random() * (variance + 1));
}

export function playerAttack(
  player: PlayerSession,
  mob: LiveMob,
  world: World,
  partyPeers: PlayerSession[] = [],
): CombatResult {
  const tmpl = world.mobs.get(mob.templateId)!;
  const messages: string[] = [];
  const dmg = rollDamage(player.totalAttack, tmpl.defense);
  mob.hp -= dmg;
  messages.push(`You strike ${tmpl.name} for ${dmg} damage! (${mob.hp}/${mob.maxHp} HP)`);

  if (mob.hp <= 0) {
    return handleMobDeath(player, mob, world, messages, partyPeers);
  }

  const counterDmg = rollDamage(tmpl.attack, player.totalDefense);
  player.data.hp -= counterDmg;
  messages.push(`${tmpl.name} hits you for ${counterDmg} damage! (${player.data.hp}/${player.data.maxHp} HP)`);

  if (player.data.hp <= 0) {
    player.data.hp = 0;
    return { messages, playerDied: true, mobKilled: false };
  }

  return { messages, playerDied: false, mobKilled: false };
}

function resolveAbility(player: PlayerSession, slot: 1 | 2) {
  const cls = CLASSES[player.data.className];
  if (slot === 2) {
    if (player.data.level < cls.ability2Level) {
      return { name: cls.ability2, damage: 0, cost: 0, locked: true, level: cls.ability2Level };
    }
    return { name: cls.ability2, damage: cls.ability2Damage, cost: cls.ability2Cost, locked: false, level: cls.ability2Level };
  }
  return { name: cls.ability, damage: cls.abilityDamage, cost: cls.abilityCost, locked: false, level: 1 };
}

export function playerAbility(
  player: PlayerSession,
  mob: LiveMob,
  world: World,
  slot: 1 | 2 = 1,
): CombatResult {
  const ability = resolveAbility(player, slot);
  const messages: string[] = [];

  if (ability.locked) {
    messages.push(`Ability unlocks at level ${ability.level}.`);
    return { messages, playerDied: false, mobKilled: false };
  }

  if (player.data.mp < ability.cost) {
    messages.push(`Not enough MP! ${ability.name} costs ${ability.cost} MP.`);
    return { messages, playerDied: false, mobKilled: false };
  }

  player.data.mp -= ability.cost;
  const tmpl = world.mobs.get(mob.templateId)!;
  const dmg = rollDamage(ability.damage + player.totalAttack, tmpl.defense);
  mob.hp -= dmg;
  messages.push(`*** ${ability.name}! *** You deal ${dmg} damage to ${tmpl.name}! (${mob.hp}/${mob.maxHp} HP)`);

  if (mob.hp <= 0) {
    return handleMobDeath(player, mob, world, messages, []);
  }

  const counterDmg = rollDamage(tmpl.attack, player.totalDefense);
  player.data.hp -= counterDmg;
  messages.push(`${tmpl.name} retaliates for ${counterDmg} damage! (${player.data.hp}/${player.data.maxHp} HP)`);

  if (player.data.hp <= 0) {
    player.data.hp = 0;
    return { messages, playerDied: true, mobKilled: false };
  }

  return { messages, playerDied: false, mobKilled: false };
}

export function playerAbilityVsPlayer(
  attacker: PlayerSession,
  defender: PlayerSession,
  slot: 1 | 2 = 1,
): PvpRoundResult {
  const ability = resolveAbility(attacker, slot);
  const attackerMessages: string[] = [];
  const defenderMessages: string[] = [];

  if (ability.locked) {
    return { attackerMessages: [`Ability unlocks at level ${ability.level}.`], defenderMessages: [], defenderDied: false, attackerDied: false };
  }
  if (attacker.data.mp < ability.cost) {
    return { attackerMessages: [`Not enough MP! ${ability.name} costs ${ability.cost} MP.`], defenderMessages: [], defenderDied: false, attackerDied: false };
  }

  attacker.data.mp -= ability.cost;
  const dmg = rollDamage(ability.damage + attacker.totalAttack, defender.totalDefense);
  defender.data.hp -= dmg;
  attackerMessages.push(`*** ${ability.name}! *** You deal ${dmg} damage to ${defender.username}! (${defender.data.hp}/${defender.data.maxHp} HP)`);
  defenderMessages.push(`*** ${attacker.username} uses ${ability.name}! *** You take ${dmg} damage! (${defender.data.hp}/${defender.data.maxHp} HP)`);

  if (defender.data.hp <= 0) {
    defender.data.hp = 0;
    return { attackerMessages, defenderMessages, defenderDied: true, attackerDied: false };
  }

  const counterDmg = rollDamage(defender.totalAttack, attacker.totalDefense);
  attacker.data.hp -= counterDmg;
  attackerMessages.push(`${defender.username} retaliates for ${counterDmg} damage! (${attacker.data.hp}/${attacker.data.maxHp} HP)`);
  defenderMessages.push(`You retaliate for ${counterDmg} damage! (${attacker.data.hp}/${attacker.data.maxHp} HP)`);

  if (attacker.data.hp <= 0) {
    attacker.data.hp = 0;
    return { attackerMessages, defenderMessages, defenderDied: false, attackerDied: true };
  }

  return { attackerMessages, defenderMessages, defenderDied: false, attackerDied: false };
}

function handleMobDeath(
  player: PlayerSession,
  mob: LiveMob,
  world: World,
  messages: string[],
  partyPeers: PlayerSession[] = [],
): CombatResult {
  const tmpl = world.mobs.get(mob.templateId)!;
  world.removeMob(player.roomId, mob.instanceId);
  player.combatTarget = null;
  player.pvpTarget = null;

  messages.push(`*** ${tmpl.name} has been slain! ***`);
  const gold = tmpl.gold.min + Math.floor(Math.random() * (tmpl.gold.max - tmpl.gold.min + 1));
  if (gold > 0) {
    player.data.gold += gold;
    messages.push(`You loot ${gold} gold.`);
  }

  messages.push(...player.addXp(tmpl.xp));
  if (partyPeers.length > 0) {
    const share = Math.floor(tmpl.xp / partyPeers.length);
    for (const peer of partyPeers) {
      if (peer.username === player.username) continue;
      const peerMsgs = peer.addXp(share);
      messages.push(`[Party] ${peer.username} gains ${share} XP.`);
      void peerMsgs;
    }
  }

  if (tmpl.loot) {
    for (const drop of tmpl.loot) {
      if (Math.random() < drop.chance) {
        player.addItem(drop.itemId);
        messages.push(`You found: ${ITEMS[drop.itemId]?.name ?? drop.itemId}`);
      }
    }
  }

  for (const qp of player.getActiveQuests()) {
    const quest = world.quests.get(qp.questId);
    if (!quest) continue;
    for (const obj of quest.objectives) {
      if (obj.type === 'kill' && obj.target === mob.templateId) {
        qp.progress[obj.target] = (qp.progress[obj.target] ?? 0) + 1;
      }
      if (obj.type === 'collect') {
        qp.progress[obj.target] = player.countItem(obj.target);
      }
    }
    if (checkQuestComplete(quest, qp)) {
      qp.status = 'complete';
      messages.push(`Quest complete: ${quest.name}! Return to the quest giver.`);
    }
  }

  return { messages, playerDied: false, mobKilled: true };
}

export function playerAttackPlayer(
  attacker: PlayerSession,
  defender: PlayerSession,
): PvpRoundResult {
  const attackerMessages: string[] = [];
  const defenderMessages: string[] = [];

  const dmg = rollDamage(attacker.totalAttack, defender.totalDefense);
  defender.data.hp -= dmg;
  attackerMessages.push(
    `You strike ${defender.username} for ${dmg} damage! (${defender.data.hp}/${defender.data.maxHp} HP)`,
  );
  defenderMessages.push(
    `${attacker.username} strikes you for ${dmg} damage! (${defender.data.hp}/${defender.data.maxHp} HP)`,
  );

  if (defender.data.hp <= 0) {
    defender.data.hp = 0;
    return { attackerMessages, defenderMessages, defenderDied: true, attackerDied: false };
  }

  const counterDmg = rollDamage(defender.totalAttack, attacker.totalDefense);
  attacker.data.hp -= counterDmg;
  attackerMessages.push(
    `${defender.username} hits you for ${counterDmg} damage! (${attacker.data.hp}/${attacker.data.maxHp} HP)`,
  );
  defenderMessages.push(
    `You hit ${attacker.username} for ${counterDmg} damage! (${attacker.data.hp}/${attacker.data.maxHp} HP)`,
  );

  if (attacker.data.hp <= 0) {
    attacker.data.hp = 0;
    return { attackerMessages, defenderMessages, defenderDied: false, attackerDied: true };
  }

  return { attackerMessages, defenderMessages, defenderDied: false, attackerDied: false };
}

export function clearPvpBetween(
  a: PlayerSession,
  b: PlayerSession | null,
): void {
  a.clearCombat();
  if (b) b.clearCombat();
}

export function pvpVictoryXp(victimLevel: number): number {
  return Math.max(10, victimLevel * 15);
}

export function handlePlayerDeath(player: PlayerSession, killer?: string): string[] {
  player.clearCombat();
  player.data.hp = Math.floor(player.data.maxHp * 0.5);
  player.data.gold = Math.max(0, player.data.gold - Math.floor(player.data.gold * 0.1));
  player.data.roomId = 'eldermoor_square';
  const lines = [
    killer
      ? `*** You have been slain by ${killer}! ***`
      : '*** You have been defeated! ***',
    'You awaken in Eldermoor Town Square, battered but alive.',
    `You lost some gold. HP restored to ${player.data.hp}.`,
  ];
  return lines;
}

File: crates/realm-core/src/combat.rs
use crate::items::{format_item_name, ITEMS};
use crate::player::PlayerSession;
use crate::quests::{check_quest_complete, QuestStatus};
use crate::types::class_stats;
use crate::world::{LiveMob, World};
use rand::Rng;
use std::collections::HashSet;
use std::sync::LazyLock;

pub static SAFE_ZONES: LazyLock<HashSet<&'static str>> =
    LazyLock::new(|| HashSet::from(["Eldermoor"]));

pub fn is_safe_zone(world: &World, room_id: &str) -> bool {
    SAFE_ZONES.contains(world.get_zone(room_id).as_str())
}

#[derive(Debug, Clone)]
pub struct CombatResult {
    pub messages: Vec<String>,
    pub player_died: bool,
    pub mob_killed: bool,
}

#[derive(Debug, Clone)]
pub struct PvpRoundResult {
    pub attacker_messages: Vec<String>,
    pub defender_messages: Vec<String>,
    pub defender_died: bool,
    pub attacker_died: bool,
}

struct ResolvedAbility {
    name: &'static str,
    damage: i32,
    cost: i32,
    locked: bool,
    level: u32,
}

fn roll_damage(attack: i32, defense: i32) -> i32 {
    let base = (attack - defense / 2).max(1);
    let variance = (base as f64 * 0.3).floor() as i32;
    let mut rng = rand::thread_rng();
    base + rng.gen_range(0..=variance)
}

fn resolve_ability(player: &PlayerSession, slot: u8) -> ResolvedAbility {
    let cls = class_stats(player.data.class_name);
    if slot == 2 {
        if player.data.level < cls.ability2_level {
            return ResolvedAbility {
                name: cls.ability2,
                damage: 0,
                cost: 0,
                locked: true,
                level: cls.ability2_level,
            };
        }
        return ResolvedAbility {
            name: cls.ability2,
            damage: cls.ability2_damage,
            cost: cls.ability2_cost,
            locked: false,
            level: cls.ability2_level,
        };
    }
    ResolvedAbility {
        name: cls.ability,
        damage: cls.ability_damage,
        cost: cls.ability_cost,
        locked: false,
        level: 1,
    }
}

pub fn player_attack(
    player: &mut PlayerSession,
    mob: &mut LiveMob,
    world: &World,
    party_peers: &mut [&mut PlayerSession],
) -> CombatResult {
    let tmpl = world
        .mobs()
        .get(&mob.template_id)
        .expect("mob template missing")
        .clone();
    let mut messages = Vec::new();
    let dmg = roll_damage(player.total_attack(), tmpl.defense);
    mob.hp -= dmg;
    messages.push(format!(
        "You strike {} for {dmg} damage! ({}/{}) HP)",
        tmpl.name, mob.hp, mob.max_hp
    ));

    if mob.hp <= 0 {
        return handle_mob_death(player, mob, world, messages, party_peers);
    }

    let counter_dmg = roll_damage(tmpl.attack, player.total_defense());
    player.data.hp -= counter_dmg;
    messages.push(format!(
        "{} hits you for {counter_dmg} damage! ({}/{}) HP)",
        tmpl.name, player.data.hp, player.data.max_hp
    ));

    if player.data.hp <= 0 {
        player.data.hp = 0;
        return CombatResult {
            messages,
            player_died: true,
            mob_killed: false,
        };
    }

    CombatResult {
        messages,
        player_died: false,
        mob_killed: false,
    }
}

pub fn player_ability(
    player: &mut PlayerSession,
    mob: &mut LiveMob,
    world: &World,
    slot: u8,
) -> CombatResult {
    let ability = resolve_ability(player, slot);
    let mut messages = Vec::new();

    if ability.locked {
        messages.push(format!("Ability unlocks at level {}.", ability.level));
        return CombatResult {
            messages,
            player_died: false,
            mob_killed: false,
        };
    }

    if player.data.mp < ability.cost {
        messages.push(format!(
            "Not enough MP! {} costs {} MP.",
            ability.name, ability.cost
        ));
        return CombatResult {
            messages,
            player_died: false,
            mob_killed: false,
        };
    }

    player.data.mp -= ability.cost;
    let tmpl = world
        .mobs()
        .get(&mob.template_id)
        .expect("mob template missing")
        .clone();
    let dmg = roll_damage(ability.damage + player.total_attack(), tmpl.defense);
    mob.hp -= dmg;
    messages.push(format!(
        "*** {}! *** You deal {dmg} damage to {}! ({}/{}) HP)",
        ability.name, tmpl.name, mob.hp, mob.max_hp
    ));

    if mob.hp <= 0 {
        return handle_mob_death(player, mob, world, messages, &mut []);
    }

    let counter_dmg = roll_damage(tmpl.attack, player.total_defense());
    player.data.hp -= counter_dmg;
    messages.push(format!(
        "{} retaliates for {counter_dmg} damage! ({}/{}) HP)",
        tmpl.name, player.data.hp, player.data.max_hp
    ));

    if player.data.hp <= 0 {
        player.data.hp = 0;
        return CombatResult {
            messages,
            player_died: true,
            mob_killed: false,
        };
    }

    CombatResult {
        messages,
        player_died: false,
        mob_killed: false,
    }
}

pub fn player_ability_vs_player(
    attacker: &mut PlayerSession,
    defender: &mut PlayerSession,
    slot: u8,
) -> PvpRoundResult {
    let ability = resolve_ability(attacker, slot);
    let mut attacker_messages = Vec::new();
    let mut defender_messages = Vec::new();

    if ability.locked {
        return PvpRoundResult {
            attacker_messages: vec![format!("Ability unlocks at level {}.", ability.level)],
            defender_messages: Vec::new(),
            defender_died: false,
            attacker_died: false,
        };
    }
    if attacker.data.mp < ability.cost {
        return PvpRoundResult {
            attacker_messages: vec![format!(
                "Not enough MP! {} costs {} MP.",
                ability.name, ability.cost
            )],
            defender_messages: Vec::new(),
            defender_died: false,
            attacker_died: false,
        };
    }

    attacker.data.mp -= ability.cost;
    let dmg = roll_damage(
        ability.damage + attacker.total_attack(),
        defender.total_defense(),
    );
    defender.data.hp -= dmg;
    attacker_messages.push(format!(
        "*** {}! *** You deal {dmg} damage to {}! ({}/{}) HP)",
        ability.name, defender.username(), defender.data.hp, defender.data.max_hp
    ));
    defender_messages.push(format!(
        "*** {} uses {}! *** You take {dmg} damage! ({}/{}) HP)",
        attacker.username(),
        ability.name,
        defender.data.hp,
        defender.data.max_hp
    ));

    if defender.data.hp <= 0 {
        defender.data.hp = 0;
        return PvpRoundResult {
            attacker_messages,
            defender_messages,
            defender_died: true,
            attacker_died: false,
        };
    }

    let counter_dmg = roll_damage(defender.total_attack(), attacker.total_defense());
    attacker.data.hp -= counter_dmg;
    attacker_messages.push(format!(
        "{} retaliates for {counter_dmg} damage! ({}/{}) HP)",
        defender.username(),
        attacker.data.hp,
        attacker.data.max_hp
    ));
    defender_messages.push(format!(
        "You retaliate for {counter_dmg} damage! ({}/{}) HP)",
        attacker.data.hp,
        attacker.data.max_hp
    ));

    if attacker.data.hp <= 0 {
        attacker.data.hp = 0;
        return PvpRoundResult {
            attacker_messages,
            defender_messages,
            defender_died: false,
            attacker_died: true,
        };
    }

    PvpRoundResult {
        attacker_messages,
        defender_messages,
        defender_died: false,
        attacker_died: false,
    }
}

fn handle_mob_death(
    player: &mut PlayerSession,
    mob: &LiveMob,
    world: &World,
    mut messages: Vec<String>,
    party_peers: &mut [&mut PlayerSession],
) -> CombatResult {
    let tmpl = world
        .mobs()
        .get(&mob.template_id)
        .expect("mob template missing")
        .clone();
    world.remove_mob(player.room_id(), &mob.instance_id);
    player.clear_combat();

    let elite_tag = if mob.elite.unwrap_or(false) || tmpl.elite {
        " ***ELITE***"
    } else {
        ""
    };
    let boss_tag = if tmpl.boss { " ***BOSS***" } else { "" };
    messages.push(format!(
        "*** {}{}{} has been slain! ***",
        tmpl.name, elite_tag, boss_tag
    ));
    player.data.kills += 1;
    if mob.template_id.contains("goblin") {
        player.data.goblin_kills += 1;
    }

    let xp_mult = if mob.elite.unwrap_or(false) || tmpl.elite {
        1.5
    } else if tmpl.boss {
        2.0
    } else {
        1.0
    };
    let gold_range = tmpl.gold.max - tmpl.gold.min + 1;
    let mut rng = rand::thread_rng();
    let gold = if gold_range > 0 {
        tmpl.gold.min + rng.gen_range(0..gold_range)
    } else {
        tmpl.gold.min
    };
    if gold > 0 {
        player.data.gold += gold;
        messages.push(format!("You loot {gold} gold."));
    }

    let xp_gain = (tmpl.xp as f64 * xp_mult).floor() as i32;
    messages.extend(player.add_xp(xp_gain));

    if !party_peers.is_empty() {
        let share = tmpl.xp / party_peers.len() as i32;
        for peer in party_peers.iter_mut() {
            if peer.username() == player.username() {
                continue;
            }
            let _ = peer.add_xp(share);
            messages.push(format!("[Party] {} gains {share} XP.", peer.username()));
        }
    }

    for drop in &tmpl.loot {
        if rng.gen::<f64>() < drop.chance {
            player.add_item(&drop.item_id);
            let _ = ITEMS.get(&drop.item_id);
            messages.push(format!("You found: {}", format_item_name(&drop.item_id)));
        }
    }

    let active_quest_ids: Vec<String> = player
        .get_active_quests()
        .into_iter()
        .map(|q| q.quest_id.clone())
        .collect();

    for quest_id in active_quest_ids {
        let Some(quest) = world.quests().get(&quest_id).cloned() else {
            continue;
        };
        let collect_counts: Vec<(String, i32)> = quest
            .objectives
            .iter()
            .filter(|obj| obj.objective_type == "collect")
            .map(|obj| (obj.target.clone(), player.count_item(&obj.target)))
            .collect();

        if let Some(qp) = player
            .data
            .quests
            .iter_mut()
            .find(|q| q.quest_id == quest_id && q.status == QuestStatus::Active)
        {
            for obj in &quest.objectives {
                if obj.objective_type == "kill" && obj.target == mob.template_id {
                    *qp.progress.entry(obj.target.clone()).or_insert(0) += 1;
                }
            }
            for (target, count) in collect_counts {
                qp.progress.insert(target, count);
            }
            if check_quest_complete(&quest, qp) {
                qp.status = QuestStatus::Complete;
                messages.push(format!(
                    "Quest complete: {}! Return to the quest giver.",
                    quest.name
                ));
            }
        }
    }

    CombatResult {
        messages,
        player_died: false,
        mob_killed: true,
    }
}

pub fn player_attack_player(
    attacker: &mut PlayerSession,
    defender: &mut PlayerSession,
) -> PvpRoundResult {
    let mut attacker_messages = Vec::new();
    let mut defender_messages = Vec::new();

    let dmg = roll_damage(attacker.total_attack(), defender.total_defense());
    defender.data.hp -= dmg;
    attacker_messages.push(format!(
        "You strike {} for {dmg} damage! ({}/{}) HP)",
        defender.username(),
        defender.data.hp,
        defender.data.max_hp
    ));
    defender_messages.push(format!(
        "{} strikes you for {dmg} damage! ({}/{}) HP)",
        attacker.username(),
        defender.data.hp,
        defender.data.max_hp
    ));

    if defender.data.hp <= 0 {
        defender.data.hp = 0;
        return PvpRoundResult {
            attacker_messages,
            defender_messages,
            defender_died: true,
            attacker_died: false,
        };
    }

    let counter_dmg = roll_damage(defender.total_attack(), attacker.total_defense());
    attacker.data.hp -= counter_dmg;
    attacker_messages.push(format!(
        "{} hits you for {counter_dmg} damage! ({}/{}) HP)",
        defender.username(),
        attacker.data.hp,
        attacker.data.max_hp
    ));
    defender_messages.push(format!(
        "You hit {} for {counter_dmg} damage! ({}/{}) HP)",
        attacker.username(),
        attacker.data.hp,
        attacker.data.max_hp
    ));

    if attacker.data.hp <= 0 {
        attacker.data.hp = 0;
        return PvpRoundResult {
            attacker_messages,
            defender_messages,
            defender_died: false,
            attacker_died: true,
        };
    }

    PvpRoundResult {
        attacker_messages,
        defender_messages,
        defender_died: false,
        attacker_died: false,
    }
}

pub fn clear_pvp_between(a: &mut PlayerSession, b: Option<&mut PlayerSession>) {
    a.clear_combat();
    if let Some(b) = b {
        b.clear_combat();
    }
}

pub fn pvp_victory_xp(victim_level: u32) -> i32 {
    (victim_level as i32 * 15).max(10)
}

pub fn player_death(player: &mut PlayerSession, killer: Option<&str>) -> Vec<String> {
    handle_player_death(player, killer)
}

pub fn handle_player_death(player: &mut PlayerSession, killer: Option<&str>) -> Vec<String> {
    player.data.deaths += 1;
    player.clear_combat();
    player.data.hp = (player.data.max_hp as f64 * 0.5).floor() as i32;
    player.data.gold = (player.data.gold - (player.data.gold as f64 * 0.1).floor() as i32).max(0);
    player.data.room_id = "eldermoor_square".into();

    vec![
        if let Some(killer) = killer {
            format!("*** You have been slain by {killer}! ***")
        } else {
            "*** You have been defeated! ***".into()
        },
        "You awaken in Eldermoor Town Square, battered but alive.".into(),
        format!("You lost some gold. HP restored to {}.", player.data.hp),
    ]
}

pub fn player_attack_pvp(
    attacker: &mut PlayerSession,
    defender: &mut PlayerSession,
) -> PvpRoundResult {
    player_attack_player(attacker, defender)
}

pub fn player_cast(
    player: &mut PlayerSession,
    mob: &mut LiveMob,
    world: &World,
) -> CombatResult {
    player_ability(player, mob, world, 1)
}

pub fn player_use_ability(
    player: &mut PlayerSession,
    mob: &mut LiveMob,
    world: &World,
) -> CombatResult {
    player_ability(player, mob, world, 1)
}

pub fn player_use_ability2(
    player: &mut PlayerSession,
    mob: &mut LiveMob,
    world: &World,
) -> CombatResult {
    player_ability(player, mob, world, 2)
}

pub fn flee_combat(player: &mut PlayerSession) -> Vec<String> {
    player.clear_combat();
    vec!["You flee from combat!".into()]
}